crates/turbo-tasks-fetch/src/lib.rs

```diff
@@ -39,13 +46,118 @@ pub async fn fetch(url: StringVc, user_agent: OptionStringVc) -> Result<HttpResp
         builder = builder.header("User-Agent", user_agent);
     }
 
-    let response = builder.send().await?;
-    let status = response.status().as_u16();
-    let body = response.bytes().await?.to_vec();
+    let response = builder.send().await;
+    match response {
+        Ok(response) => {
+            let status = response.status().as_u16();
+            let body = response.bytes().await?.to_vec();
 
-    Ok(HttpResponse {
-        status,
-        body: HttpResponseBodyVc::cell(HttpResponseBody(body)),
+            Ok(HttpResponse {
+                status,
+                body: HttpResponseBodyVc::cell(HttpResponseBody(body)),
+            }
+            .into())
+        }
+        Err(err) => {
+            fetch_issue_from_reqwest_error(&err, context)
+                .cell()
+                .as_issue()
+                .emit();
+            Err(err.into())
+        }
+    }
+}
+
+#[turbo_tasks::value(shared)]
+enum FetchIssueKind {
+    Connect,
+    Timeout,
+    Status(u16),
+    Other,
+}
+
+#[turbo_tasks::value(shared)]
+struct FetchIssue {
+    pub context: FileSystemPathVc,
+    pub url: OptionStringVc,
+    pub kind: FetchIssueKindVc,
+    pub detail: StringVc,
+}
+
+fn fetch_issue_from_reqwest_error(
+    error: &reqwest::Error,
+    issue_context: FileSystemPathVc,
+) -> FetchIssue {
+    let kind = if error.is_connect() {
+        FetchIssueKind::Connect
+    } else if error.is_timeout() {
+        FetchIssueKind::Timeout
+    } else if let Some(status) = error.status() {
+        FetchIssueKind::Status(status.as_u16())
+    } else {
+        FetchIssueKind::Other
+    };
+
+    FetchIssue {
+        context: issue_context,
+        detail: StringVc::cell(error.to_string()),
+        url: OptionStringVc::cell(error.url().map(|url| url.to_string())),
```

Comment:
From the description of [`Error::url`](https://docs.rs/reqwest/latest/reqwest/struct.Error.html#method.url), I think we should use the original URL instead (and clone it if we have to).